README.md
<h1 align="center" style="border-bottom: none;">🚀 topgun-socket</h1>
<h3 align="center">Pub/sub and RPC toolkit for <a href="https://github.com/TopGunBuild/topgun">TopGun</a></h3>

<p align="center">
  <a href="https://github.com/semantic-release/semantic-release">
      <img alt="semantic-release" src="https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg">
  </a>
  <a href="https://npm.im/topgun-socket">
    <img alt="npm" src="https://badgen.net/npm/v/topgun-socket">
  </a>
  <a href="https://bundlephobia.com/result?p=topgun-socket">
    <img alt="bundlephobia" src="https://img.shields.io/bundlephobia/minzip/topgun-socket.svg">
  </a>
  <a href="https://opensource.org/licenses/MIT">
      <img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg">
  </a>
  <a href="https://snyk.io/test/github/TopGunBuild/topgun">
        <img alt="Snyk" src="https://snyk.io/test/github/TopGunBuild/topgun/badge.svg">
    </a>
</p>

## Client Setting up

To install this module:

```bash
npm install topgun-socket
```

## How to use client module

The socket-client script is called `client.js`(ESM) or `client.global.js`(iife) (located in the dist directory).
Embed it in your HTML page like this:

```html
<script type="text/javascript" src="https://unpkg.com/topgun-socket@latest/dist/client.global.js"></script>
```

\* Note that the src attribute may be different depending on how you setup your HTTP server.

Once you have embedded the client `client.global.js` into your page, you will gain access to a global `TopGunSocket` object.
You may also use CommonJS `require` or ES6 module imports.

### Connect to a server

```js
let socket = TopGunSocket.create({
    hostname: "localhost",
    port: 8000,
});
```

### Transmit data

```js
// Transmit some data to the server.
// It does not expect a response from the server.
// From the server socket, it can be handled using either:
// - for await (let data of socket.receiver('foo')) {}
// - let data = await socket.receiver('foo').once()
socket.transmit('foo', 123);
```

### Invoke an RPC

```js
(async () => {

  // Invoke an RPC on the server.
  // It expects a response from the server.
  // From the server socket, it can be handled using either:
  // - for await (let req of socket.procedure('myProc')) {}
  // - let req = await socket.procedure('myProc').once()
  let result = await socket.invoke('myProc', 123);

})();
```

### Subscribe to a channel

```js
(async () => {

  // Subscribe to a channel.
  let myChannel = socket.subscribe('myChannel');

  await myChannel.listener('subscribe').once();
  // myChannel.state is now 'subscribed'.

})();
```

### Get a channel without subscribing

```js
(async () => {

  let myChannel = socket.channel('myChannel');

  // Can subscribe to the channel later as a separate step.
  myChannel.subscribe();
  await myChannel.listener('subscribe').once();
  // myChannel.state is now 'subscribed'.

})();
```

### Publish data to a channel

```js
// Publish data to the channel.
myChannel.publish('This is a message');

// Publish data to the channel from the socket.
socket.publish('myChannel', 'This is a message');
```

### Consume data from a channel

```js
(async () => {

  for await (let data of myChannel) {
    // ...
  }

})();
```

### Connect over HTTPS:

```js
let options = {
  hostname: 'securedomain.com',
  secure: true,
  port: 443,
  rejectUnauthorized: false // Only necessary during debug if using a self-signed certificate
};
// Initiate the connection to the server
let socket = TopGunSocket.create(options);
```

### Connect Options

```js
let options = {
  path: '/topgunsocket/',
  port: 8000,
  hostname: '127.0.0.1',
  autoConnect: true,
  secure: false,
  rejectUnauthorized: false,
  connectTimeout: 10000, //milliseconds
  ackTimeout: 10000, //milliseconds
  channelPrefix: null,
  disconnectOnUnload: true,
  multiplex: true,
  autoReconnectOptions: {
    initialDelay: 10000, //milliseconds
    randomness: 10000, //milliseconds
    multiplier: 1.5, //decimal
    maxDelay: 60000 //milliseconds
  },
  authEngine: null,
  codecEngine: null,
  subscriptionRetryOptions: {},
  query: {
    yourparam: 'hello'
  }
};
```

For more detailed examples of how to use topgun-socket, see `test` folder.

## How to use server module

You need to attach it to an existing Node.js http or https server (example):
```js
const http = require('http');
const topGunSocketServer = require("topgun-socket/server");

let httpServer = http.createServer();
let agServer = topGunSocketServer.attach(httpServer);

(async () => {
  // Handle new inbound sockets.
  for await (let {socket} of agServer.listener('connection')) {

    (async () => {
      // Set up a loop to handle and respond to RPCs for a procedure.
      for await (let req of socket.procedure('customProc')) {
        if (req.data.bad) {
          let error = new Error('Server failed to execute the procedure');
          error.name = 'BadCustomError';
          req.error(error);
        } else {
          req.end('Success');
        }
      }
    })();

    (async () => {
      // Set up a loop to handle remote transmitted events.
      for await (let data of socket.receiver('customRemoteEvent')) {
        // ...
      }
    })();

  }
})();

httpServer.listen(8000);
```

## Running the tests

-   Clone this repo: `git clone git@github.com:TopGunBuild/topgun-socket.git`
-   Navigate to project directory: `cd topgun-socket`
-   Install all dependencies: `npm install`
-   Run the tests: `npm test`

## Benefits of async `Iterable` over `EventEmitter`

-   **More readable**: Code is written sequentially from top to bottom. It avoids event handler callback hell. It's also much easier to write and read complex integration test scenarios.
-   **More succinct**: Event streams can be easily chained, filtered and combined using a declarative syntax (e.g. using async generators).
-   **More manageable**: No need to remember to unbind listeners with `removeListener(...)`; just `break` out of the `for-await-of` loop to stop consuming. This also encourages a more declarative style of coding which reduces the likelihood of memory leaks and unintended side effects.
-   **Less error-prone**: Each event/RPC/message can be processed sequentially in the same order that they were sent without missing any data; even if asynchronous calls are made inside middleware or listeners. On the other hand, with `EventEmitter`, the listener function for the same event cannot be prevented from running multiple times in parallel; also, asynchronous calls within middleware and listeners can affect the final order of actions; all this can cause unintended side effects.

## Reference

The code in this repository is based on `socketcluster-server` (https://github.com/SocketCluster/socketcluster-server) and `socketcluster-client` (https://github.com/SocketCluster/socketcluster-client). The code is written in typescript and is ready to run in a serverless environment.

## License

(The MIT License)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## [1.4.9](https://github.com/topgunjs/topgun-socket/compare/v1.4.8...v1.4.9) (2023-06-29)


### Bug Fixes

* add missing packages to export ([7c7e08a](https://github.com/topgunjs/topgun-socket/commit/7c7e08a943c1fe333d8379f48a34b09e9531e239))
* build packages ([7cae744](https://github.com/topgunjs/topgun-socket/commit/7cae744d117889cbf8caf5931fa2b0f10b9ced94))
* possible handle socket connection directly ([c9cbeb9](https://github.com/topgunjs/topgun-socket/commit/c9cbeb9a2ddf33eb3408bb75161e17ad15e45bf2))
* refactor code for compatibility with topgun-async-stream-emitter ([0cb7aeb](https://github.com/topgunjs/topgun-socket/commit/0cb7aeb7549b6745ad15e44c019a306bd852268c))
* **tools:** update package list ([f68230b](https://github.com/topgunjs/topgun-socket/commit/f68230bf5c47949d10b75a9bc03b68e2221b13e8))
* use WSServer only in nodejs ([0195b66](https://github.com/topgunjs/topgun-socket/commit/0195b668a17df999fd713bdcf03a3ca1f47f92c6))

## [1.4.8](https://github.com/topgunjs/topgun-socket/compare/v1.4.7...v1.4.8) (2023-06-28)


### Bug Fixes

* install topgun-async-stream-emitter, remove local async-stream-emitter ([9c5ad84](https://github.com/topgunjs/topgun-socket/commit/9c5ad8424779bee48bc9fd81930ed9f57d164fa7))

## [1.4.7](https://github.com/topgunjs/topgun-socket/compare/v1.4.6...v1.4.7) (2023-06-01)


### Bug Fixes

* build packages ([49a9b89](https://github.com/topgunjs/topgun-socket/commit/49a9b8986f5d0c4fbcda7b345889fa5be4bf6c92))

## [1.4.6](https://github.com/topgunjs/topgun-socket/compare/v1.4.5...v1.4.6) (2023-06-01)


### Bug Fixes

* **tools:** update package list ([46d9241](https://github.com/topgunjs/topgun-socket/commit/46d9241d1ca0736c37fdd575b3fdd9e28ae5e621))

## [1.4.5](https://github.com/topgunjs/topgun-socket/compare/v1.4.4...v1.4.5) (2023-06-01)


### Bug Fixes

* add missing packages to export ([28b0f50](https://github.com/topgunjs/topgun-socket/commit/28b0f501cc050b34101c906af9f1589c781377c4))

## [1.4.4](https://github.com/topgunjs/topgun-socket/compare/v1.4.3...v1.4.4) (2023-05-29)


### Bug Fixes

* use WSServer only in nodejs ([4ee532d](https://github.com/topgunjs/topgun-socket/commit/4ee532d5b0bccedb73607834c84197bb66ef2e27))

## [1.4.3](https://github.com/topgunjs/topgun-socket/compare/v1.4.2...v1.4.3) (2023-05-29)


### Bug Fixes

* possible handle socket connection directly ([f907895](https://github.com/topgunjs/topgun-socket/commit/f9078957ca9af4c3e2b073ad760907ace69c09fe))

## [1.4.2](https://github.com/topgunjs/topgun-socket/compare/v1.4.1...v1.4.2) (2023-05-28)


### Bug Fixes

* **server:** set missing typings ([284ef7a](https://github.com/topgunjs/topgun-socket/commit/284ef7ad4436bb57354959e298d82e2d96683d5a))

## [1.4.1](https://github.com/topgunjs/topgun-socket/compare/v1.4.0...v1.4.1) (2023-05-28)


### Bug Fixes

* **server:** set missing typings ([c7c9637](https://github.com/topgunjs/topgun-socket/commit/c7c96375e2f9f8e02470608dd50b12991b6b899b))

# [1.4.0](https://github.com/topgunjs/topgun-socket/compare/v1.3.3...v1.4.0) (2023-05-27)


### Bug Fixes

* fix encode for serverless ([335a78c](https://github.com/topgunjs/topgun-socket/commit/335a78cd147b7810f27784eb6b09e0a00c9b5c33))
* fix global ([f6b26f8](https://github.com/topgunjs/topgun-socket/commit/f6b26f8d907a87e0120e199a80f8a1f2a5f19cdf))


### Features

* rename paths and keys ([e38afe4](https://github.com/topgunjs/topgun-socket/commit/e38afe4ad4b63b297fe39d1a11c8f03cbac85730))

## [1.3.3](https://github.com/topgunjs/topgun-socket/compare/v1.3.2...v1.3.3) (2023-05-03)


### Bug Fixes

* upgrade topgun-buffer 1.0.5 -> 1.0.6 ([298d105](https://github.com/topgunjs/topgun-socket/commit/298d10540ea7882aa56ed4fb1a8ce3f9edce7b3c))

## [1.3.2](https://github.com/topgunjs/topgun-socket/compare/v1.3.1...v1.3.2) (2023-05-03)


### Bug Fixes

* upgrade topgun-buffer 1.0.3 -> 1.0.5 ([ecc323b](https://github.com/topgunjs/topgun-socket/commit/ecc323b5f5443920e30315e5b7fc5f37f59bbe89))

## [1.3.1](https://github.com/topgunjs/topgun-socket/compare/v1.3.0...v1.3.1) (2023-05-01)


### Bug Fixes

* fix listen function ([175a853](https://github.com/topgunjs/topgun-socket/commit/175a853bee636378939a4dbd88763ce5671c7f75))

# [1.3.0](https://github.com/topgunjs/topgun-socket/compare/v1.2.4...v1.3.0) (2023-05-01)


### Features

* **socket-server:** remove `http` param from `listen`. add typings to `attach` ([516cc12](https://github.com/topgunjs/topgun-socket/commit/516cc124248ef928ef0b8f30a97dcd3af33276f5))


### Reverts

* remove "esModuleInterop" from tsconfig ([c7bd460](https://github.com/topgunjs/topgun-socket/commit/c7bd46047b08967f2a74eb506d4a514615f1cb6b))

## [1.2.4](https://github.com/topgunjs/topgun-socket/compare/v1.2.3...v1.2.4) (2023-04-23)


### Bug Fixes

* bug fixes and improvements ([2ffdf23](https://github.com/topgunjs/topgun-socket/commit/2ffdf23a83deacdf9f60b26d8bc404b06f83c7c0))
* fix global, add test ([0968fac](https://github.com/topgunjs/topgun-socket/commit/0968facaa199bcc707380061ac9ad305755684dd))
* fix test passing ([9554836](https://github.com/topgunjs/topgun-socket/commit/9554836f61900a51ece23d45890c66e9322fa747))

server/server.ts
import { AsyncStreamEmitter } from '../async-stream-emitter';
import {
    AGServerOptions,
    AuthEngineType,
    CodecEngine,
    MIDDLEWARE_HANDSHAKE,
    MIDDLEWARE_INBOUND_RAW,
    MIDDLEWARE_INBOUND,
    MIDDLEWARE_OUTBOUND,
    outboundMiddlewareFunction,
    inboundMiddlewareFunction,
    inboundRawMiddlewareFunction,
    handshakeMiddlewareFunction,
    Middlewares,
    AuthStateChangeData,
    AuthenticationData,
    DeauthenticationData,
    BadSocketAuthTokenData,
    ConnectionData,
    SubscriptionData,
    UnsubscriptionData,
    ConnectionAbortData,
    DisconnectionData,
    ClosureData
} from './types';
import { AGServerSocket } from './server-socket';
import { SimpleExchange } from '../ag-simple-broker/simple-exchange';
import { JwtAlgorithm, JwtVerifyOptions, Secret } from '../jwt';
import { AGSimpleBroker } from '../ag-simple-broker/simple-broker';
import {
    InvalidActionError,
    InvalidArgumentsError,
    InvalidOptionsError,
    ServerProtocolError, SilentMiddlewareBlockedError
} from '../sc-errors/errors';
import { randomBytes } from '../crypto/crypto';
import { AuthEngine } from '../ag-auth';
import { formatter } from '../sc-formatter';
import { isNode } from '../utils/is-node';
import { generateId } from '../utils/generate-id';
import { AGAction } from './action';
import { WritableConsumableStream } from '../writable-consumable-stream';
import { ConsumableStream } from '../consumable-stream';

export class AGServer extends AsyncStreamEmitter<any>
{
    static MIDDLEWARE_HANDSHAKE: Middlewares   = MIDDLEWARE_HANDSHAKE;
    static MIDDLEWARE_INBOUND_RAW: Middlewares = MIDDLEWARE_INBOUND_RAW;
    static MIDDLEWARE_INBOUND: Middlewares     = MIDDLEWARE_INBOUND;
    static MIDDLEWARE_OUTBOUND: Middlewares    = MIDDLEWARE_OUTBOUND;
    static SYMBOL_MIDDLEWARE_HANDSHAKE_STREAM  = Symbol('handshakeStream');

    options: AGServerOptions;
    origins: string;
    ackTimeout: number;
    handshakeTimeout: number;
    pingInterval: number;
    pingTimeout: number;
    pingTimeoutDisabled: boolean;
    allowClientPublish: boolean;
    perMessageDeflate?: boolean|{};
    httpServer: any;
    socketChannelLimit?: number;
    protocolVersion: 1|2;
    strictHandshake: boolean;
    brokerEngine: AGSimpleBroker;
    middlewareEmitFailures: boolean;
    isReady: boolean;
    signatureKey?: Secret;
    verificationKey?: Secret;
    defaultVerificationOptions: JwtVerifyOptions|JwtAlgorithm;
    defaultSignatureOptions: {
        expiresIn: number;
        algorithm?: string;
    };
    exchange: SimpleExchange;

    clients: {
        [id: string]: AGServerSocket;
    };
    clientsCount: number;

    pendingClients: {
        [id: string]: AGServerSocket;
    };
    pendingClientsCount: number;

    auth: AuthEngineType;
    wsServer: any;
    codec: CodecEngine;
    private readonly _middleware: {[key: string]: any};
    private readonly _allowAllOrigins: boolean;
    private readonly _path: string;

    /**
     * Constructor
     */
    constructor(options?: AGServerOptions)
    {
        super();

        let opts = {
            brokerEngine            : new AGSimpleBroker(),
            wsEngine                : 'ws',
            wsEngineServerOptions   : {},
            maxPayload              : null,
            allowClientPublish      : true,
            ackTimeout              : 10000,
            handshakeTimeout        : 10000,
            strictHandshake         : true,
            pingTimeout             : 20000,
            pingTimeoutDisabled     : false,
            pingInterval            : 8000,
            origins                 : '*:*',
            path                    : '/topgunsocket/',
            protocolVersion         : 2,
            authDefaultExpiry       : 86400,
            batchOnHandshake        : false,
            batchOnHandshakeDuration: 400,
            batchInterval           : 50,
            middlewareEmitFailures  : true,
            socketStreamCleanupMode : 'kill',
            cloneData               : false,
            isNode                  : isNode()
        };

        this.options = Object.assign(opts, options);

        this._middleware = {};

        this.origins          = this.options.origins;
        this._allowAllOrigins = this.origins.indexOf('*:*') !== -1;

        this.ackTimeout          = this.options.ackTimeout;
        this.handshakeTimeout    = this.options.handshakeTimeout;
        this.pingInterval        = this.options.pingInterval;
        this.pingTimeout         = this.options.pingTimeout;
        this.pingTimeoutDisabled = this.options.pingTimeoutDisabled;
        this.allowClientPublish  = this.options.allowClientPublish;
        this.perMessageDeflate   = this.options.perMessageDeflate;
        this.httpServer          = this.options.httpServer;
        this.socketChannelLimit  = this.options.socketChannelLimit;
        this.protocolVersion     = this.options.protocolVersion;
        this.strictHandshake     = this.options.strictHandshake;

        this.brokerEngine           = this.options.brokerEngine;
        this.middlewareEmitFailures = this.options.middlewareEmitFailures;

        this._path = opts.path;

        (async () =>
        {
            for await (let { error } of this.brokerEngine.listener('error'))
            {
                this.emitWarning(error);
            }
        })();

        if (this.brokerEngine.isReady)
        {
            this.isReady = true;
            this.emit('ready', {});
        }
        else
        {
            this.isReady = false;
            (async () =>
            {
                await this.brokerEngine.listener('ready').once();
                this.isReady = true;
                this.emit('ready', {});
            })();
        }

        if (!this.options.wsEngine)
        {
            throw new InvalidOptionsError(
                'The wsEngine option must be a path or module name which points ' +
                'to a valid WebSocket engine module with a compatible interface'
            );
        }
        let WSServer = this.options.wsEngine;

        if (this.options.authPrivateKey != null || this.options.authPublicKey != null)
        {
            if (this.options.authPrivateKey == null)
            {
                throw new InvalidOptionsError(
                    'The authPrivateKey option must be specified if authPublicKey is specified'
                );
            }
            else if (this.options.authPublicKey == null)
            {
                throw new InvalidOptionsError(
                    'The authPublicKey option must be specified if authPrivateKey is specified'
                );
            }
            this.signatureKey    = this.options.authPrivateKey;
            this.verificationKey = this.options.authPublicKey;
        }
        else
        {
            if (this.options.authKey == null)
            {
                this.options.authKey = randomBytes(32).toString();
            }
            this.signatureKey    = this.options.authKey;
            this.verificationKey = this.options.authKey;
        }

        this.defaultVerificationOptions = {};
        if (this.options.authVerifyAlgorithms != null)
        {
            this.defaultVerificationOptions.algorithms = this.options.authVerifyAlgorithms;
        }
        else if (this.options.authAlgorithm != null)
        {
            this.defaultVerificationOptions.algorithms = [this.options.authAlgorithm];
        }

        this.defaultSignatureOptions = {
            expiresIn: this.options.authDefaultExpiry
        };
        if (this.options.authAlgorithm != null)
        {
            this.defaultSignatureOptions.algorithm = this.options.authAlgorithm;
        }

        if (this.options.authEngine)
        {
            this.auth = this.options.authEngine;
        }
        else
        {
            // Default authentication engine
            this.auth = new AuthEngine();
        }

        if (this.options.codecEngine)
        {
            this.codec = this.options.codecEngine;
        }
        else
        {
            // Default codec engine
            this.codec = formatter;
        }
        this.brokerEngine.setCodecEngine(this.codec);
        this.exchange = this.brokerEngine.exchange();

        this.clients      = {};
        this.clientsCount = 0;

        this.pendingClients      = {};
        this.pendingClientsCount = 0;

        let wsServerOptions          = this.options.wsEngineServerOptions || {};
        wsServerOptions.server       = this.httpServer;
        wsServerOptions.verifyClient = this.verifyHandshake.bind(this);

        if (wsServerOptions.path == null && this._path != null)
        {
            wsServerOptions.path = this._path;
        }
        if (wsServerOptions.perMessageDeflate == null && this.perMessageDeflate != null)
        {
            wsServerOptions.perMessageDeflate = this.perMessageDeflate;
        }
        if (wsServerOptions.handleProtocols == null && this.options.handleProtocols != null)
        {
            wsServerOptions.handleProtocols = this.options.handleProtocols;
        }
        if (wsServerOptions.maxPayload == null && opts.maxPayload != null)
        {
            wsServerOptions.maxPayload = opts.maxPayload;
        }
        if (wsServerOptions.clientTracking == null)
        {
            wsServerOptions.clientTracking = false;
        }

        if (this.options.isNode)
        {
            // @ts-ignore
            this.wsServer = new WSServer(wsServerOptions);

            this.wsServer.on('error', this._handleServerError.bind(this));
            this.wsServer.on('connection', this._handleSocketConnection.bind(this));
        }
        else
        {
            this.wsServer = WSServer;
            this.wsServer.addEventListener('close', this._closeOrErrorHandler.bind(this));
            this.wsServer.addEventListener('error', this._closeOrErrorHandler.bind(this));
        }
    }

    // -----------------------------------------------------------------------------------------------------
    // @ Public methods
    // -----------------------------------------------------------------------------------------------------

    setAuthEngine(authEngine: AuthEngineType): void
    {
        this.auth = authEngine;
    }

    setCodecEngine(codecEngine: CodecEngine): void
    {
        this.codec = codecEngine;
        this.brokerEngine.setCodecEngine(codecEngine);
    }

    emit(eventName: 'error', data: {error: Error}): void;
    emit(eventName: 'warning', data: {warning: Error}): void;
    emit(eventName: 'handshake', data: {socket: AGServerSocket}): void;
    emit(eventName: 'authenticationStateChange', data: AuthStateChangeData): void;
    emit(eventName: 'authentication', data: AuthenticationData): void;
    emit(eventName: 'deauthentication', data: DeauthenticationData): void;
    emit(eventName: 'badSocketAuthToken', data: BadSocketAuthTokenData): void;
    emit(eventName: 'connection', data: ConnectionData): void;
    emit(eventName: 'subscription', data: SubscriptionData): void;
    emit(eventName: 'unsubscription', data: UnsubscriptionData): void;
    emit(eventName: 'connectionAbort', data: ConnectionAbortData): void;
    emit(eventName: 'disconnection', data: DisconnectionData): void;
    emit(eventName: 'closure', data: ClosureData): void;
    emit(eventName: 'ready', data: any): void;
    emit(eventName: string, data: any): void
    {
        return super.emit(eventName, data);
    }

    listener(eventName: 'error'): ConsumableStream<{error: Error}>;
    listener(eventName: 'warning'): ConsumableStream<{warning: Error}>;
    listener(eventName: 'handshake'): ConsumableStream<{socket: AGServerSocket}>;
    listener(eventName: 'authenticationStateChange'): ConsumableStream<AuthStateChangeData>;
    listener(eventName: 'authentication'): ConsumableStream<AuthenticationData>;
    listener(eventName: 'deauthentication'): ConsumableStream<DeauthenticationData>;
    listener(eventName: 'badSocketAuthToken'): ConsumableStream<BadSocketAuthTokenData>;
    listener(eventName: 'connection'): ConsumableStream<ConnectionData>;
    listener(eventName: 'subscription'): ConsumableStream<SubscriptionData>;
    listener(eventName: 'unsubscription'): ConsumableStream<UnsubscriptionData>;
    listener(eventName: 'connectionAbort'): ConsumableStream<ConnectionAbortData>;
    listener(eventName: 'disconnection'): ConsumableStream<DisconnectionData>;
    listener(eventName: 'closure'): ConsumableStream<ClosureData>;
    listener(eventName: 'ready'): ConsumableStream<any>
    listener(eventName: string): ConsumableStream<any>
    {
        return super.listener(eventName);
    }

    emitError(error: Error): void
    {
        this.emit('error', { error });
    }

    emitWarning(warning: Error): void
    {
        this.emit('warning', { warning });
    }

    close(keepSocketsOpen?: boolean): Promise<void>
    {
        this.isReady = false;
        return new Promise((resolve, reject) =>
        {
            this.wsServer.close((err) =>
            {
                if (err)
                {
                    reject(err);
                    return;
                }
                resolve();
            });
            if (!keepSocketsOpen)
            {
                for (let socket of Object.values(this.clients))
                {
                    socket.terminate();
                }
            }
        });
    }

    getPath(): string
    {
        return this._path;
    }

    generateId(): string
    {
        return generateId();
    }

    setMiddleware(type: typeof MIDDLEWARE_HANDSHAKE, middleware: handshakeMiddlewareFunction): void;
    setMiddleware(type: typeof MIDDLEWARE_INBOUND_RAW, middleware: inboundRawMiddlewareFunction): void;
    setMiddleware(type: typeof MIDDLEWARE_INBOUND, middleware: inboundMiddlewareFunction): void;
    setMiddleware(type: typeof MIDDLEWARE_OUTBOUND, middleware: outboundMiddlewareFunction): void
    setMiddleware(type: Middlewares, middleware: any)
    {
        if (
            type !== AGServer.MIDDLEWARE_HANDSHAKE &&
            type !== AGServer.MIDDLEWARE_INBOUND_RAW &&
            type !== AGServer.MIDDLEWARE_INBOUND &&
            type !== AGServer.MIDDLEWARE_OUTBOUND
        )
        {
            throw new InvalidArgumentsError(
                `Middleware type "${type}" is not supported`
            );
        }
        if (this._middleware[type])
        {
            throw new InvalidActionError(`Middleware type "${type}" has already been set`);
        }
        this._middleware[type] = middleware;
    }

    removeMiddleware(type: Middlewares): void
    {
        delete this._middleware[type];
    }

    hasMiddleware(type: Middlewares): boolean
    {
        return !!this._middleware[type];
    }

    async verifyHandshake(info, callback): Promise<any>
    {
        let req    = info.req;
        let origin = info.origin;
        if (origin === 'null' || origin == null)
        {
            origin = '*';
        }
        let ok: boolean|number = false;

        if (this._allowAllOrigins)
        {
            ok = true;
        }
        else
        {
            try
            {
                const parser = new URL(origin);
                const port   = parser.port || (parser.protocol === 'https:' ? 443 : 80);
                ok           = ~this.origins.indexOf(parser.hostname + ':' + port) ||
                    ~this.origins.indexOf(parser.hostname + ':*') ||
                    ~this.origins.indexOf('*:' + port);
            }
            catch (e)
            {
            }
        }

        let middlewareHandshakeStream  = new WritableConsumableStream();
        middlewareHandshakeStream.type = AGServer.MIDDLEWARE_HANDSHAKE;

        req[AGServer.SYMBOL_MIDDLEWARE_HANDSHAKE_STREAM] = middlewareHandshakeStream;

        let handshakeMiddleware = this._middleware[AGServer.MIDDLEWARE_HANDSHAKE];
        if (handshakeMiddleware)
        {
            handshakeMiddleware(middlewareHandshakeStream);
        }

        let action     = new AGAction();
        action.request = req;
        action.type    = AGAction.HANDSHAKE_WS;

        try
        {
            await this.processMiddlewareAction(middlewareHandshakeStream, action);
        }
        catch (error)
        {
            middlewareHandshakeStream.close();
            callback(false, 401, typeof error === 'string' ? error : error.message);
            return;
        }

        if (ok)
        {
            callback(true);
            return;
        }

        let error = new ServerProtocolError(
            `Failed to authorize socket handshake - Invalid origin: ${origin}`
        );
        this.emitWarning(error);

        middlewareHandshakeStream.close();
        callback(false, 403, error.message);
    }

    async processMiddlewareAction(
        middlewareStream,
        action: AGAction,
        socket?
    ): Promise<{data: any, options: any}>
    {
        if (!this.hasMiddleware(middlewareStream.type))
        {
            return { data: action.data, options: null };
        }
        middlewareStream.write(action);

        let newData;
        let options = null;
        try
        {
            let result = await action.promise;
            if (result)
            {
                newData = result.data;
                options = result.options;
            }
        }
        catch (error)
        {
            let clientError;
            if (!error)
            {
                error       = new SilentMiddlewareBlockedError(
                    `The ${action.type} AGAction was blocked by ${middlewareStream.type} middleware`,
                    middlewareStream.type
                );
                clientError = error;
            }
            else if (error.silent)
            {
                clientError = new SilentMiddlewareBlockedError(
                    `The ${action.type} AGAction was blocked by ${middlewareStream.type} middleware`,
                    middlewareStream.type
                );
            }
            else
            {
                clientError = error;
            }
            if (this.middlewareEmitFailures)
            {
                if (socket)
                {
                    socket.emitError(error);
                }
                else
                {
                    this.emitWarning(error);
                }
            }
            throw clientError;
        }

        if (newData === undefined)
        {
            newData = action.data;
        }

        return { data: newData, options };
    }

    // -----------------------------------------------------------------------------------------------------
    // @ Private methods
    // -----------------------------------------------------------------------------------------------------

    private _closeOrErrorHandler(error?: Error): void
    {
        if (error)
        {
            this.emitError(error);
        }
        this.close();
    }

    private _handleServerError(error: Error): void
    {
        if (typeof error === 'string')
        {
            error = new ServerProtocolError(error);
        }
        this.emitError(error);
    }

    private _handleSocketConnection(wsSocket, upgradeReq): void
    {
        if (!wsSocket.upgradeReq)
        {
            // Normalize ws modules to match.
            wsSocket.upgradeReq = upgradeReq;
        }

        const socketId    = this.generateId();
        const agSocket    = new AGServerSocket(socketId, this, wsSocket, this.protocolVersion);
        agSocket.exchange = this.exchange;

        let inboundRawMiddleware = this._middleware[AGServer.MIDDLEWARE_INBOUND_RAW];
        if (inboundRawMiddleware)
        {
            inboundRawMiddleware(agSocket.middlewareInboundRawStream);
        }

        let inboundMiddleware = this._middleware[AGServer.MIDDLEWARE_INBOUND];
        if (inboundMiddleware)
        {
            inboundMiddleware(agSocket.middlewareInboundStream);
        }

        let outboundMiddleware = this._middleware[AGServer.MIDDLEWARE_OUTBOUND];
        if (outboundMiddleware)
        {
            outboundMiddleware(agSocket.middlewareOutboundStream);
        }

        // Emit event to signal that a socket handshake has been initiated.
        this.emit('handshake', { socket: agSocket });
    }
}